src/puppetlabs/puppetdb/query_eng/engine.clj

```diff
@@ -1539,6 +1574,39 @@
 
     annotated-query))
 
+;; Top-level parsing
+
+(def experimental-entities
+  #{:event-counts :aggregate-event-counts})
+
+(defn warn-experimental
+  "Show a warning if the endpoint is experimental."
+  [entity]
+  (when (contains? experimental-entities entity)
+    (log/warn (format
+                "The %s entity is experimental and may be altered or removed in the future."
+                (name entity)))))
+
+(defn parse-query-context
```

Comment:
This magic extracts the 'from' and returns the remaining query and entity. In a way its a compatibility layer, later on we want to push the 'from' further down into the query engine, but munging still happens at the query_engine.clj level, so we still need this extraction for now ...